### ISSUE TITLE
[On device learning] Use model.fit() when pretraining in VGG16

### ISSUE THREAD

**Issue by DonghakPark** [tensorflow/tensorflow#56668]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Support

### Source

source

### Tensorflow Version

tf2.9

### Custom Code

Yes

### OS Platform and Distribution

Linux Ubuntu 20.04

### Python version

3.9

### CUDA/cuDNN version

11.7

### GPU model and memory

GTX 1650

### Current Behaviour?

This is version info
```
2022-07-04 13:54:25.617651: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
v2.9.0-18-gd8ce9f9c301 2.9.1
```

According to This Example (https://www.tensorflow.org/lite/examples/on_device_training/overview)
using tf_function custom defined train for pretrain

I tried to Train VGG16 and CIFAR100 while conducting the test, but the training was not proceed,

- It works, but loss and accuracy do not change

--> so I am going to proceed with pretrain using model.fit as below, is this supported by On_device_Learning? <Pretrain seems like Working Well>

--> Or what action should I take if I try to use Gradient Tape in VGG16? (EveryThing Works Well in Lenet5)

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf
from tensorflow.keras import initializers
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

image_shape = (32,32,3)
num_classes = 100

class Model(tf.Module):
    
    def __init__(self):
        
        self.model = tf.keras.Sequential([
        tf.keras.Input(shape=image_shape, name='input_img', dtype = tf.float32),
                
        tf.keras.layers.Conv2D(64, (3,3), padding='same', activation='relu',bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(64, (3,3), padding='same', activation='relu',bias_initializer=initializers.Zeros()),
        tf.keras.layers.MaxPooling2D(pool_size=(2,2), padding='same'),
        
        tf.keras.layers.Conv2D(128, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(128, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.MaxPooling2D(pool_size=(2,2), padding='same'),
        
        tf.keras.layers.Conv2D(256, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(256, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(256, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.MaxPooling2D(pool_size=(2,2), padding='same'),
        
        tf.keras.layers.Conv2D(512, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(512, (3,3), padding='same', activation='relu',bias_initializer=initializers.Zeros() ),
        tf.keras.layers.Conv2D(512, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.MaxPooling2D(pool_size=(2,2), padding='same'),
        
        tf.keras.layers.Conv2D(512, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(512, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.Conv2D(512, (3,3), padding='same', activation='relu' ,bias_initializer=initializers.Zeros()),
        tf.keras.layers.MaxPooling2D(pool_size=(2,2), padding='same'),

        tf.keras.layers.Flatten(),
        tf.keras.layers.Dense(512,bias_initializer=initializers.Zeros(), activation = 'relu'),
        tf.keras.layers.BatchNormalization(),
        tf.keras.layers.Dense(100,bias_initializer=initializers.Zeros(), activation = 'softmax'),
        ])

        self.model.compile(
            optimizer = 'sgd',
            loss = tf.keras.losses.CategoricalCrossentropy(from_logits=False),
            metrics=['accuracy']
        )
        
    @tf.function(input_signature=[
        tf.TensorSpec([None, image_shape[0], image_shape[1], image_shape[2]], tf.float32),
        tf.TensorSpec([None, 100], tf.float32),
    ])
    def train(self, x, y):
        with tf.GradientTape() as tape:
            prediction = self.model(x) # prediction 
            loss = self.model.loss(y, prediction) #calculate loss
        
        gradients = tape.gradient(loss, self.model.trainable_variables) # calculate gradient
        self.model.optimizer.apply_gradients(zip(gradients, self.model.trainable_variables)) #backpropagation
        
        result = {"loss": loss, "prediction": prediction}
        
        return result

    
    @tf.function(input_signature=[
        tf.TensorSpec([None, image_shape[0], image_shape[1], image_shape[2]], tf.float32),    ])
    def infer(self, x):
        logits = self.model(x)
        probabilities = tf.nn.softmax(logits, axis=-1)
        return {
            "output": probabilities,
            "logits": logits
        }

    @tf.function(input_signature=[tf.TensorSpec(shape=[], dtype=tf.string)])
    def save(self, checkpoint_path):
        tensor_names = [weight.name for weight in self.model.weights]
        tensors_to_save = [weight.read_value() for weight in self.model.weights]
        tf.raw_ops.Save(
            filename=checkpoint_path, tensor_names=tensor_names,
            data=tensors_to_save, name='save')
        return {
            "checkpoint_path": checkpoint_path
        }

    @tf.function(input_signature=[tf.TensorSpec(shape=[], dtype=tf.string)])
    def restore(self, checkpoint_path):
        restored_tensors = {}
        for var in self.model.weights:
            restored = tf.raw_ops.Restore(
                file_pattern=checkpoint_path, tensor_name=var.name, dt=var.dtype,
                name='restore')
            var.assign(restored)
            restored_tensors[var.name] = restored
        return restored_tensors




m = Model()

(x_train, y_train), (x_test, y_test) = tf.keras.datasets.cifar100.load_data()

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')

y_train = tf.keras.utils.to_categorical(y_train, num_classes)
y_test = tf.keras.utils.to_categorical(y_test, num_classes)

train_image, val_image, train_label, val_label = train_test_split(x_train, y_train, test_size = 0.15)

history = m.model.fit(x_train, y_train, batch_size = 128, epochs = 3 )

m.save('./model.ckpt')

SAVED_MODEL_DIR = "saved_model"

tf.saved_model.save(
    m,
    SAVED_MODEL_DIR,
    signatures={
        'train':
            m.train.get_concrete_function() ,
        'infer':
            m.infer.get_concrete_function(),
        'save':
            m.save.get_concrete_function(),
        'restore':
            m.restore.get_concrete_function(),
    })

converter = tf.lite.TFLiteConverter.from_saved_model(SAVED_MODEL_DIR)
converter.target_spec.supported_ops = [
    tf.lite.OpsSet.TFLITE_BUILTINS,  # enable TensorFlow Lite ops.
    tf.lite.OpsSet.SELECT_TF_OPS  # enable TensorFlow ops.
]
converter.experimental_enable_resource_variables = True
tflite_model = converter.convert()
with open('model_vgg_fit.tflite', 'wb') as f:
    f.write(tflite_model)
```

### Relevant log output

_No response_</details>

**Comment by mohantym** [comment]
Hi @DonghakPark!
You have not [preprocessed ](https://www.tensorflow.org/lite/examples/on_device_training/overview#preprocess_the_dataset)the data properly  (Normalization and one-hot encoding missed) . Instead of training the model directly , you should call the[ train](https://www.tensorflow.org/lite/examples/on_device_training/overview#train_the_model) function in between batches of dataset to train it with recurring data.


```
train_images = (train_images / 255.0).astype(np.float32)
test_images = (test_images / 255.0).astype(np.float32)
```

Thank you!

**Comment by DonghakPark** [comment]
Hi @mohantym Thank you for answering.

I changed it as below. But use the train of tf_function instead of fit, it does not seem to be learned.
- For the CIFAR100 data, /255.0 and categorical were used.

Doesn't On Device Support VGG16? or Should I change The Method of Gradient Tape (Train Function).
--> IF you have On Device Learning Example for VGG16, Can you Share it with me ?

```python

m = Model()

(train_images, train_labels), (test_images, test_labels) = tf.keras.datasets.cifar100.load_data()

train_images = (train_images / 255.0 ).astype(np.float32)
test_images = (test_images / 255.0).astype(np.float32)

train_labels = tf.keras.utils.to_categorical(train_labels)
test_labels = tf.keras.utils.to_categorical(test_labels)

NUM_EPOCHS = 10
BATCH_SIZE = 128

print(train_images.shape, train_labels.shape)
print("-----TRIANING STRAT-----")
for i in range(NUM_EPOCHS):
    for batch_idx in range(0, 50000, BATCH_SIZE):
        
        result = m.train(train_images[batch_idx:batch_idx+BATCH_SIZE,:,:,:], train_labels[batch_idx:batch_idx+BATCH_SIZE,:])

        predict = result['prediction']
        predict = np.argmax(predict, axis = 1)
        y_answer = np.argmax(train_labels[batch_idx:batch_idx+BATCH_SIZE,:], axis=1)

    print("Train : ", i ,"/", NUM_EPOCHS, "LOSS : ", np.mean(result['loss']), "ACC : ",  accuracy_score(y_answer, predict))

```

50000 Data for each Epoch
```shell
(50000, 32, 32, 3) (50000, 100)
-----TRIANING STRAT-----
Train :  0 / 10 LOSS :  4.605494 ACC :  0.0
Train :  1 / 10 LOSS :  4.605406 ACC :  0.0375
Train :  2 / 10 LOSS :  4.6053505 ACC :  0.025
Train :  3 / 10 LOSS :  4.6052504 ACC :  0.025
Train :  4 / 10 LOSS :  4.6051908 ACC :  0.025
Train :  5 / 10 LOSS :  4.6050973 ACC :  0.025
Train :  6 / 10 LOSS :  4.605029 ACC :  0.0125
Train :  7 / 10 LOSS :  4.6049867 ACC :  0.0125
...

```

**Comment by mohantym** [comment]
Ok @DonghakPark !
Thanks for the update. @sachinprasadhs ! Could you please look at this issue. 
Thank you!

**Comment by DonghakPark** [comment]
@sachinprasadhs @Xhark is There any Updates?
